src/metatracer.rs

```diff
@@ -30,57 +30,89 @@
 // DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-use std::collections::hash_map::{Entry, HashMap};
-use std::hash::Hash;
-use std::sync::Mutex;
+use std::sync::atomic::{AtomicU32, Ordering};
 
-const HOT_THRESHOLD: u16 = 10;
+pub type HotThreshold = u32;
+const DEFAULT_HOT_THRESHOLD: HotThreshold = 50;
 
-/// The current meta-tracing state of a given location in the end-user's code.
-#[derive(Debug, PartialEq)]
-enum LocPhase {
-    /// Below `HOT_THRESHOLD`, but counting up towards it.
-    Counting(u16),
-    /// A trace is being collected.
-    Tracing,
-    /// A compiled trace is available.
-    Compiled
+// The current meta-tracing phase of a given location in the end-user's code. Consists of a tag and
+// (optionally) a value.
+const PHASE_TAG     : u32 = 0b11; // All of the other PHASE_ tags must fit in this.
+const PHASE_COUNTING: u32 = 0b00; // The value specifies the current hot count.
+const PHASE_TRACING : u32 = 0b01;
+const PHASE_COMPILED: u32 = 0b10;
+
+/// A `Location` uniquely identifies a control point position in the end-user's program (and is
+/// used by the `MetaTracer` to store data about that location). In other words, every position
+/// that can be a control point also needs to have one `Location` value associated with it, and
+/// that same `Location` value must always be used to identify that control point.
+///
+/// As this may suggest, program positions that can't be control points don't need an associated
+/// `Location`. For interpreters that can't (or don't want) to be as selective, a simple (if
+/// moderately wasteful) mechanism is for every bytecode or AST node to have its own `Location`
+/// (even for bytecodes or nodes that can't be control points).
+pub struct Location {
+    pack: AtomicU32
+}
+
+impl Location {
+    /// Create a fresh Location suitable for passing to `MetaTracer::control_point`.
+    pub fn new() -> Self {
+        Self { pack: AtomicU32::new(PHASE_COUNTING) }
+    }
 }
 
 /// A meta-tracer.
-pub struct MetaTracer<Loc> {
-    counts: Mutex<HashMap<Loc, LocPhase>>
+pub struct MetaTracer {
+    hot_threshold: AtomicU32
 }
 
-impl<Loc: Eq + Hash> MetaTracer<Loc> {
+impl MetaTracer {
     pub fn new() -> Self {
         Self {
-            counts: Mutex::new(HashMap::new())
+            hot_threshold: AtomicU32::new(DEFAULT_HOT_THRESHOLD)
         }
     }
 
+    /// Update the hot threshold for this meta-tracer instance to `threshold`.
+    pub fn set_hot_threshold(&mut self, threshold: HotThreshold) {
```

Comment:
Do we expect the user to be able to change the hot threshold after initialising the JIT?

If not, then why don't we kill this function, have the hot threshold as a parameter to `new()`, and then the hot count can be a regular `u32`, thus simplifying the code in the control point (no need for the atomic load/store dance).

Reply:
Good point. Yes, that would simplify things.

Reply:
[Well, I think the default constructor should still take no parameters; but there's no harm in having a second constructor.]